rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/Rename.rsc

```diff
@@ -229,25 +231,32 @@ Maybe[loc] rascalLocationOfName(Nonterminal nt) = just(nt.src);
 Maybe[loc] rascalLocationOfName(NonterminalLabel l) = just(l.src);
 default Maybe[loc] rascalLocationOfName(Tree t) = nothing();
 
-private tuple[set[IllegalRenameReason] reasons, list[TextEdit] edits] computeTextEdits(WorkspaceInfo ws, start[Module] m, set[RenameLocation] defs, set[RenameLocation] uses, str name) {
+private tuple[set[IllegalRenameReason] reasons, list[TextEdit] edits] computeTextEdits(WorkspaceInfo ws, start[Module] m, set[RenameLocation] defs, set[RenameLocation] uses, str name, ChangeAnnotationRegister registerChangeAnnotation) {
     if (reasons := rascalCollectIllegalRenames(ws, m, defs.l, uses.l, name), reasons != {}) {
         return <reasons, []>;
     }
 
     replaceName = rascalEscapeName(name);
 
-    set[RenameLocation] renames = defs + uses;
-    set[loc] renameLocs = renames.l;
-    map[loc, loc] namesAt = rascalFindNamesInUseDefs(m, renameLocs);
-
-    return <{}, [{just(annotation), *_} := renames[l]
-                 ? replace(namesAt[l], replaceName, annotation = annotation)
-                 : replace(namesAt[l], replaceName)
-                 | l <- renameLocs]>;
+    rel[loc l, Maybe[ChangeAnnotationId] ann, bool isDef] renames =
+        {<l, a, true>  | <l, a> <- defs}
+      + {<l, a, false> | <l, a> <- uses};
+    rel[loc name, loc useDef] nameOfUseDef = rascalFindNamesInUseDefs(m, renames.l);
+
+    ChangeAnnotationId defAnno = registerChangeAnnotation("Definitions", MANDATORY_CHANGE_DESCRIPTION, false);
+    ChangeAnnotationId useAnno = registerChangeAnnotation("References", MANDATORY_CHANGE_DESCRIPTION, false);
+
+    // Note: if the implementer of the rename logic has attached annotations to multiple rename suggestions that have the same
```

Comment:
this reads like it should be written in a different way? as in, that the comprehension below is too simple?

Reply:
Well, the we do not have that information at all here currently. For rename locations that already have an annotation added at this stage (which are only `has` uses at this point in time), we only have the id of the annotation, but not its params. That still points in the direction of a necessary refactoring, probably. But I wanted to get the fix out first.

Reply:
We could consider keeping a pair of `<loc, Maybe[ChangeAnnotation]>` instead, decouple the two using an `id` later, before returning to the Java world. Or even in Java.

Reply:
Made an issue for this: https://github.com/usethesource/rascal-language-servers/issues/527